packages/datadog-instrumentations/src/amqp10.js

```diff
@@ -0,0 +1,68 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+addHook({ name: 'amqp10', file: 'lib/sender_link.js', versions: ['>=3'] }, SenderLink => {
+  const startCh = channel('apm:amqp10:send:start')
+  const asyncEndCh = channel('apm:amqp10:send:async-end')
+  const endCh = channel('apm:amqp10:send:end')
+  const errorCh = channel('apm:amqp10:send:error')
+  shimmer.wrap(SenderLink.prototype, 'send', send => function (msg, options) {
+    if (!startCh.hasSubscribers) {
+      return send.apply(this, arguments)
+    }
+    startCh.publish({ link: this })
+    try {
+      const promise = send.apply(this, arguments)
+
+      if (!promise) {
+        finish(asyncEndCh, errorCh)
+        return promise
+      }
+
+      promise.then(() => AsyncResource.bind(finish(asyncEndCh, errorCh)),
```

Comment:
This should bind the function passed to then, not the result of `finish`. Same for the next line.

Reply:
Since a promise can only ever be either resolved or rejected itt would also be more efficient to use a single async resource instance to bind both.